internal/impl/gcp/input_bigquery_select_test.go

```diff
@@ -0,0 +1,230 @@
+	spec := newBigQuerySelectInputConfig()
+
+	parsed, err := spec.ParseYAML(testBQInputYAML, nil)
+	require.NoError(t, err)
+
+	inp, err := newBigQuerySelectInput(parsed, nil)
+	require.NoError(t, err)
+
+	msg, ack, err := inp.Read(context.Background())
+	require.ErrorIs(t, err, service.ErrNotConnected)
+	require.Nil(t, msg)
+	require.Nil(t, ack)
+}
+
+func TestGCPBigQuerySelectInput_IteratorError(t *testing.T) {
+	spec := newBigQuerySelectInputConfig()
+
+	parsed, err := spec.ParseYAML(testBQInputYAML, nil)
+	require.NoError(t, err)
+
+	inp, err := newBigQuerySelectInput(parsed, nil)
+	require.NoError(t, err)
+
+	rows := []string{
+		`{"total_count":25568,"word":"the"}`,
+	}
+
+	testErr := errors.New("simulated error")
+	inp.iterator = &testBQIterator{
+		err:     testErr,
+		rawRows: rows,
+	}
+
+	msg, ack, err := inp.Read(context.Background())
+	require.ErrorIs(t, err, testErr)
+	require.Nil(t, msg)
+	require.Nil(t, ack)
+}
+
+// mockBQQueryAPI mocks the BigQuery jobs and queries APIs in order to test
+// the Connect method of bigQuerySelectInput
+type mockBQQueryAPI struct {
+	mock.Mock
+}
+
+func (m *mockBQQueryAPI) handleListJobs(rw http.ResponseWriter, r *http.Request) {
+	args := m.Mock.Called()
+	rw.Write([]byte(args.String(0)))
+}
+
+func (m *mockBQQueryAPI) handleGetJob(rw http.ResponseWriter, r *http.Request) {
+	args := m.Mock.Called()
+	rw.Write([]byte(args.String(0)))
+}
+
+func (m *mockBQQueryAPI) handleListQueries(rw http.ResponseWriter, r *http.Request) {
+	args := m.Mock.Called()
+	rw.Write([]byte(args.String(0)))
+}
+
+func happyBQMock(t *testing.T) (*httptest.Server, *mockBQQueryAPI) {
+	mockJobResult := `{
+    "id": "test-job-id",
+    "jobReference": { "jobId": "test-job-id", "project": "EU", "projectId": "job-project" },
+    "configuration": { "jobType": "QUERY", "query": {} },
+    "status": { "state": "DONE" }
+  }`
+	mockQueryResult := `{
+    "jobReference": { "jobId": "test-job-id", "project": "EU", "projectId": "job-project" },
+    "jobComplete": true,
+    "totalRows": "10",
+    "schema": { "fields": [ { "name": "word", "type": "STRING" }, { "name": "total_count", "type": "INTEGER" } ] }
+  }`
+
+	mockAPI := &mockBQQueryAPI{}
+	mockAPI.On("handleListJobs").Return(mockJobResult)
+	mockAPI.On("handleGetJob").Return(mockJobResult)
+	mockAPI.On("handleListQueries").Return(mockQueryResult)
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/projects/job-project/jobs", mockAPI.handleListJobs)
+	mux.HandleFunc("/projects/job-project/jobs/test-job-id", mockAPI.handleGetJob)
+	mux.HandleFunc("/projects/job-project/queries/test-job-id", mockAPI.handleListQueries)
+	mux.HandleFunc("/", func(rw http.ResponseWriter, r *http.Request) {
+		t.Logf("%s %s", r.Method, r.URL)
+		panic("mock endpoint not implemented")
+	})
+
+	server := httptest.NewServer(mux)
+
+	t.Cleanup(func() {
+		server.Close()
+	})
+
+	return server, mockAPI
+}
+
+func TestGCPBigQuerySelectInput_Connect(t *testing.T) {
```

Comment:
This is the only test that needed elaborate mocking of the BigQuery API. The other tests that focus on `Read` were kept simple by passing a mock iterator to the input instance. Any future changes to the `Read` logic can continue to use this simpler testing path. Having said that, if we consider this particular test to be too much ceremony then I'm comfortable with deleting it and the associated mock server.